File: Contests/Luogu/LGR-162-Div.3/PD.py
import sys

# import math
# from bisect import *
# from collections import *
# from functools import *
# from heapq import *
# from itertools import *
# from random import *
# from string import *
# from types import GeneratorType

# region fastio
input = lambda: sys.stdin.readline().rstrip()
sint = lambda: int(input())
mint = lambda: map(int, input().split())
ints = lambda: list(map(int, input().split()))
# print = lambda d: sys.stdout.write(str(d) + "\n")
# endregion fastio

# # region interactive
# def printQry(a, b) -> None:
#     sa = str(a)
#     sb = str(b)
#     print(f"? {sa} {sb}", flush = True)

# def printAns(ans) -> None:
#     s = str(ans)
#     print(f"! {s}", flush = True)
# # endregion interactive

# # region dfsconvert
# def bootstrap(f, stack=[]):
#     def wrappedfunc(*args, **kwargs):
#         if stack:
#             return f(*args, **kwargs)
#         else:
#             to = f(*args, **kwargs)
#             while True:
#                 if type(to) is GeneratorType:
#                     stack.append(to)
#                     to = next(to)
#                 else:
#                     stack.pop()
#                     if not stack:
#                         break
#                     to = stack[-1].send(to)
#             return to
#     return wrappedfunc
# # endregion dfsconvert

# MOD = 998244353
# MOD = 10 ** 9 + 7
# DIR = ((-1, 0), (0, 1), (1, 0), (0, -1))

class RangeLink:
    def __init__(self, n: int) -> None:
        self.vals = [1] * n
        self.s = 0
        self.nxt = list(range(1, n + 1))
    
    def modify(self, l: int, r: int) -> None:
        x = l
        while x <= r:
            if self.vals[x]: self.s += 1
            self.vals[x] = 0
            self.nxt[x], x = self.nxt[r], self.nxt[x]
    
    def query(self, l: int, r: int) -> int:
        res = 0
        x = l
        while x <= r:
            res += self.vals[x]
            x = self.nxt[x]
        return res

def solve() -> None:
    n, m, k, q = mint()
    ans = [0] * k
    qry = [[] for _ in range(2)]
    for _ in range(q):
        op, l, r, c, t = mint()
        qry[t].append((op, l, r, c - 1))
        
    row, col = RangeLink(n), RangeLink(m)

    for op, l, r, c in qry[1][::-1]:
        if op == 1: # row
            rc, cc = row.query(l - 1, r - 1), col.s
            ans[c] += rc * (m - cc)
            row.modify(l - 1, r - 1)
        else:       # col
            rc, cc = row.s, col.query(l - 1, r - 1)
            ans[c] += (n - rc) * cc
            col.modify(l - 1, r - 1)

    for op, l, r, c in qry[0]:
        if op == 1: # row
            rc, cc = row.query(l - 1, r - 1), col.s
            ans[c] += rc * (m - cc)
            row.modify(l - 1, r - 1)
        else:       # col
            rc, cc = row.s, col.query(l - 1, r - 1)
            ans[c] += (n - rc) * cc
            col.modify(l - 1, r - 1)
    
    print(*ans)

solve()

File: Contests/Luogu/LGR-162-Div.3/PD.go
package main

import (
	"bufio"
	. "fmt"
	"io"
	"os"
)

func solve(_r io.Reader, _w io.Writer) {
	// in := bufio.NewReader(_r)
	out := bufio.NewWriter(_w)
	defer out.Flush()
	buf := make([]byte, 4096)
	_i := len(buf)
	rc := func() byte {
		if _i == len(buf) {
			_r.Read(buf)
			_i = 0
		}
		b := buf[_i]
		_i++
		return b
	}
	r := func() (x int) {
		b := rc()
		for ; '0' > b; b = rc() {
		}
		for ; '0' <= b; b = rc() {
			x = x*10 + int(b&15)
		}
		return
	}

	n, m, k, q := r(), r(), r(), r()
	ans := make([]int, k)

	row_vals := make([]int, n)
	col_vals := make([]int, m)
	row_sum, col_sum := 0, 0
	row_nxt := make([]int, n)
	col_nxt := make([]int, m)
	for i := 0; i < n; i++ {
		row_nxt[i] = i + 1
	}
	for i := 0; i < m; i++ {
		col_nxt[i] = i + 1
	}

	modify := func(vals, nxt []int, s *int, l, r int) {
		x := l
		for x <= r {
			if vals[x] == 0 {
				*s++
			}
			vals[x] = 1
			nxt[x], x = nxt[r], nxt[x]
		}
	}

	query := func(vals, nxt []int, l, r int) int {
		res := 0
		for x := l; x <= r; x = nxt[x] {
			res += vals[x] ^ 1
		}
		return res
	}

	qry := make([][]int, q)
	for i := 0; i < q; i++ {
		qry[i] = make([]int, 5)
		qry[i][0], qry[i][1], qry[i][2], qry[i][3], qry[i][4] = r(), r(), r(), r(), r()
		qry[i][3]--
	}

	for i := q - 1; i >= 0; i-- {
		if qry[i][4] == 1 {
			if qry[i][0] == 1 {
				rc, cc := query(row_vals, row_nxt, qry[i][1]-1, qry[i][2]-1), col_sum
				ans[qry[i][3]] += rc * (m - cc)
				modify(row_vals, row_nxt, &row_sum, qry[i][1]-1, qry[i][2]-1)
			} else {
				rc, cc := row_sum, query(col_vals, col_nxt, qry[i][1]-1, qry[i][2]-1)
				ans[qry[i][3]] += (n - rc) * cc
				modify(col_vals, col_nxt, &col_sum, qry[i][1]-1, qry[i][2]-1)
			}
		}
	}

	for i := 0; i < q; i++ {
		if qry[i][4] == 0 {
			if qry[i][0] == 1 {
				rc, cc := query(row_vals, row_nxt, qry[i][1]-1, qry[i][2]-1), col_sum
				ans[qry[i][3]] += rc * (m - cc)
				modify(row_vals, row_nxt, &row_sum, qry[i][1]-1, qry[i][2]-1)
			} else {
				rc, cc := row_sum, query(col_vals, col_nxt, qry[i][1]-1, qry[i][2]-1)
				ans[qry[i][3]] += (n - rc) * cc
				modify(col_vals, col_nxt, &col_sum, qry[i][1]-1, qry[i][2]-1)
			}
		}
	}

	for _, v := range ans {
		Fprint(out, v, " ")
	}
}

func main() { solve(os.Stdin, os.Stdout) }
